The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/22222298541/job/64280481748) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/64280481748) [(plain)](https://github.com/rust-lang/rust/commit/92b7394e31d193fcc5c0cfd5059a4e80b4fb8093/checks/64280481748/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6708 files
tidy check
tidy [deps]: Dependency for rustc-main not explicitly permitted: registry+https://github.com/rust-lang/crates.io-index#tracing-serde@0.2.0
Go to `src/tools/tidy/src/deps.rs:277` for the list.
tidy [deps]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: deps
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1612:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1365:29

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:44
  local time: Fri Feb 20 11:33:47 UTC 2026
  network time: Fri, 20 Feb 2026 11:33:47 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_log/Cargo.toml
+++ b/compiler/rustc_log/Cargo.toml
@@ -7,7 +7,7 @@ edition = "2024"
 # tidy-alphabetical-start
 tracing = "0.1.41"
 tracing-core = "0.1.34"
-tracing-subscriber = { version = "0.3.3", default-features = false, features = ["fmt", "env-filter", "smallvec", "parking_lot", "ansi"] }
+tracing-subscriber = { version = "0.3.3", default-features = false, features = ["fmt", "env-filter", "smallvec", "parking_lot", "ansi", "json"] }
 tracing-tree = "0.3.1"
 # tidy-alphabetical-end
 

--- a/compiler/rustc_log/src/lib.rs
+++ b/compiler/rustc_log/src/lib.rs
@@ -66,6 +66,7 @@ pub struct LoggerConfig {
 
 impl LoggerConfig {
     pub fn from_env(env: &str) -> Self {
+        // NOTE: documented in the dev guide. If you change this, also update it!
         LoggerConfig {
             filter: env::var(env),
             color_logs: env::var(format!("{env}_COLOR")),

--- a/compiler/rustc_log/src/lib.rs
+++ b/compiler/rustc_log/src/lib.rs
@@ -35,13 +35,16 @@
 
 use std::env::{self, VarError};
 use std::fmt::{self, Display};
+use std::fs::File;
 use std::io::{self, IsTerminal};
+use std::sync::Mutex;
 
 use tracing::dispatcher::SetGlobalDefaultError;
 use tracing::{Event, Subscriber};
 use tracing_subscriber::filter::{Directive, EnvFilter, LevelFilter};
 use tracing_subscriber::fmt::FmtContext;
 use tracing_subscriber::fmt::format::{self, FmtSpan, FormatEvent, FormatFields};
+use tracing_subscriber::fmt::writer::BoxMakeWriter;
 use tracing_subscriber::layer::SubscriberExt;
 use tracing_subscriber::{Layer, Registry};
 // Re-export tracing
@@ -56,6 +59,7 @@ pub struct LoggerConfig {
     pub verbose_thread_ids: Result<String, VarError>,
     pub backtrace: Result<String, VarError>,
     pub json: Result<String, VarError>,
+    pub output_target: Result<String, VarError>,
     pub wraptree: Result<String, VarError>,
     pub lines: Result<String, VarError>,
 }
@@ -71,6 +75,7 @@ impl LoggerConfig {
             wraptree: env::var(format!("{env}_WRAPTREE")),
             lines: env::var(format!("{env}_LINES")),
             json: env::var(format!("{env}_FORMAT_JSON")),
+            output_target: env::var(format!("{env}_OUTPUT_TARGET")),
         }
     }
 }
@@ -143,6 +148,17 @@ where
         Err(_) => false,
     };
 
+    let output_target: BoxMakeWriter = match cfg.output_target {
+        Ok(v) => match File::options().create(true).write(true).truncate(true).open(&v) {
+            Ok(i) => BoxMakeWriter::new(Mutex::new(i)),
+            Err(e) => {
+                eprintln!("couldn't open {v} as a log target: {e:?}");
+                BoxMakeWriter::new(io::stderr)
+            }
+        },
+        Err(_) => BoxMakeWriter::new(io::stderr),
+    };
+
     let layer = if json {
         let format = tracing_subscriber::fmt::format()
             .json()
@@ -152,7 +168,7 @@ where
         let fmt_layer = tracing_subscriber::fmt::layer()
             .json()
             .event_format(format)
-            .with_writer(io::stderr)
+            .with_writer(output_target)
             .with_target(true)
             .with_ansi(false)
             .with_thread_ids(verbose_thread_ids)
@@ -161,7 +177,7 @@ where
         Layer::boxed(fmt_layer)
     } else {
         let mut layer = tracing_tree::HierarchicalLayer::default()
-            .with_writer(io::stderr)
+            .with_writer(output_target)
             .with_ansi(color_logs)
             .with_targets(true)
             .with_verbose_exit(verbose_entry_exit)

--- a/compiler/rustc_log/src/lib.rs
+++ b/compiler/rustc_log/src/lib.rs
@@ -39,11 +39,11 @@ use std::io::{self, IsTerminal};
 
 use tracing::dispatcher::SetGlobalDefaultError;
 use tracing::{Event, Subscriber};
-use tracing_subscriber::Registry;
 use tracing_subscriber::filter::{Directive, EnvFilter, LevelFilter};
 use tracing_subscriber::fmt::FmtContext;
-use tracing_subscriber::fmt::format::{self, FormatEvent, FormatFields};
+use tracing_subscriber::fmt::format::{self, FmtSpan, FormatEvent, FormatFields};
 use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::{Layer, Registry};
 // Re-export tracing
 pub use {tracing, tracing_core, tracing_subscriber};
 
@@ -55,6 +55,7 @@ pub struct LoggerConfig {
     pub verbose_entry_exit: Result<String, VarError>,
     pub verbose_thread_ids: Result<String, VarError>,
     pub backtrace: Result<String, VarError>,
+    pub json: Result<String, VarError>,
     pub wraptree: Result<String, VarError>,
     pub lines: Result<String, VarError>,
 }
@@ -69,6 +70,7 @@ impl LoggerConfig {
             backtrace: env::var(format!("{env}_BACKTRACE")),
             wraptree: env::var(format!("{env}_WRAPTREE")),
             lines: env::var(format!("{env}_LINES")),
+            json: env::var(format!("{env}_FORMAT_JSON")),
         }
     }
 }
@@ -136,23 +138,48 @@ where
         Err(_) => false,
     };
 
-    let mut layer = tracing_tree::HierarchicalLayer::default()
-        .with_writer(io::stderr)
-        .with_ansi(color_logs)
-        .with_targets(true)
-        .with_verbose_exit(verbose_entry_exit)
-        .with_verbose_entry(verbose_entry_exit)
-        .with_indent_amount(2)
-        .with_indent_lines(lines)
-        .with_thread_ids(verbose_thread_ids)
-        .with_thread_names(verbose_thread_ids);
+    let json = match cfg.json {
+        Ok(v) => &v == "1",
+        Err(_) => false,
+    };
 
-    if let Ok(v) = cfg.wraptree {
-        match v.parse::<usize>() {
-            Ok(v) => layer = layer.with_wraparound(v),
-            Err(_) => return Err(Error::InvalidWraptree(v)),
+    let layer = if json {
+        let format = tracing_subscriber::fmt::format()
+            .json()
+            .with_span_list(true)
+            .with_source_location(true);
+
+        let fmt_layer = tracing_subscriber::fmt::layer()
+            .json()
+            .event_format(format)
+            .with_writer(io::stderr)
+            .with_target(true)
+            .with_ansi(false)
+            .with_thread_ids(verbose_thread_ids)
+            .with_thread_names(verbose_thread_ids)
+            .with_span_events(FmtSpan::ACTIVE);
+        Layer::boxed(fmt_layer)
+    } else {
+        let mut layer = tracing_tree::HierarchicalLayer::default()
+            .with_writer(io::stderr)
+            .with_ansi(color_logs)
+            .with_targets(true)
+            .with_verbose_exit(verbose_entry_exit)
+            .with_verbose_entry(verbose_entry_exit)
+            .with_indent_amount(2)
+            .with_indent_lines(lines)
+            .with_thread_ids(verbose_thread_ids)
+            .with_thread_names(verbose_thread_ids);
+
+        if let Ok(v) = cfg.wraptree {
+            match v.parse::<usize>() {
+                Ok(v) => layer = layer.with_wraparound(v),
+                Err(_) => return Err(Error::InvalidWraptree(v)),
+            }
         }
-    }
+
+        Layer::boxed(layer)
+    };
 
     let subscriber = build_subscriber();
     // NOTE: It is important to make sure that the filter is applied on the last layer

--- a/src/doc/rustc-dev-guide/src/tracing.md
+++ b/src/doc/rustc-dev-guide/src/tracing.md
@@ -12,6 +12,25 @@ To see the logs, you need to set the `RUSTC_LOG` environment variable to your lo
 The full syntax of the log filters can be found in the [rustdoc
 of `tracing-subscriber`](https://docs.rs/tracing-subscriber/0.2.24/tracing_subscriber/filter/struct.EnvFilter.html#directives).
 
+## Environment variables
+
+This is an overview of the environment variables rustc accepts to customize
+its tracing output. The definition of these can mostly be found in `compiler/rustc_log/src/lib.rs`.
+
+|  Name                     | Usage                                                                                                                   |
+| ------------------------- | ----------------------------------------------------------------------------------------------------------------------- |
+| `RUSTC_LOG`               | Tracing filters (see the rest of this page)                                                                             |
+| `RUSTC_LOG_COLOR`         | `always`, `never` or `auto`.                                                                                            |
+| `RUSTC_LOG_ENTRY_EXIT`    | if set and not '0', log span entry/exit.                                                                                |
+| `RUSTC_LOG_THREAD_IDS`    | if set and equal to '1', also log thread ids.                                                                           |
+| `RUSTC_LOG_BACKTRACE`     | Capture and print a backtrace if a trace is emitted with a target that matches the provided string value.               |
+| `RUSTC_LOG_LINES`         | If `1`, indents log lines based on depth.                                                                               |
+| `RUSTC_LOG_FORMAT_JSON`   | If `1`, outputs logs as JSON. The exact parameters can be found in `rustc_log/src/lib.rs` but the format is *UNSTABLE*. |
+| `RUSTC_LOG_OUTPUT_TARGET` | If provided, logs go to the provided file name instead of stderr.                                                       |
+
+
+
+
 ## Function level filters
 
 Lots of functions in rustc are annotated with

--- a/src/tools/tidy/src/deps.rs
+++ b/src/tools/tidy/src/deps.rs
@@ -450,6 +450,7 @@ const PERMITTED_RUSTC_DEPENDENCIES: &[&str] = &[
     "tracing-attributes",
     "tracing-core",
     "tracing-log",
+    "tracing-serde",
     "tracing-subscriber",
     "tracing-tree",
     "twox-hash",

